projects/coreutils-9.5/man/du.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.48.5.
.TH DU "1" "March 2024" "GNU coreutils 9.5" "User Commands"
.SH NAME
du \- estimate file space usage
.SH SYNOPSIS
.B du
[\fI\,OPTION\/\fR]... [\fI\,FILE\/\fR]...
.br
.B du
[\fI\,OPTION\/\fR]... \fI\,--files0-from=F\/\fR
.SH DESCRIPTION
.\" Add any additional description here
.PP
Summarize device usage of the set of FILEs, recursively for directories.
.PP
Mandatory arguments to long options are mandatory for short options too.
.TP
\fB\-0\fR, \fB\-\-null\fR
end each output line with NUL, not newline
.TP
\fB\-a\fR, \fB\-\-all\fR
write counts for all files, not just directories
.TP
\fB\-\-apparent\-size\fR
print apparent sizes rather than device usage; although
the apparent size is usually smaller, it may be
larger due to holes in ('sparse') files, internal
fragmentation, indirect blocks, and the like
.TP
\fB\-B\fR, \fB\-\-block\-size\fR=\fI\,SIZE\/\fR
scale sizes by SIZE before printing them; e.g.,
\&'\-BM' prints sizes in units of 1,048,576 bytes;
see SIZE format below
.TP
\fB\-b\fR, \fB\-\-bytes\fR
equivalent to '\-\-apparent\-size \fB\-\-block\-size\fR=\fI\,1\/\fR'
.TP
\fB\-c\fR, \fB\-\-total\fR
produce a grand total
.TP
\fB\-D\fR, \fB\-\-dereference\-args\fR
dereference only symlinks that are listed on the
command line
.TP
\fB\-d\fR, \fB\-\-max\-depth\fR=\fI\,N\/\fR
print the total for a directory (or file, with \fB\-\-all\fR)
only if it is N or fewer levels below the command
line argument;  \fB\-\-max\-depth\fR=\fI\,0\/\fR is the same as
\fB\-\-summarize\fR
.TP
\fB\-\-files0\-from\fR=\fI\,F\/\fR
summarize device usage of the
NUL\-terminated file names specified in file F;
if F is \-, then read names from standard input
.TP
\fB\-H\fR
equivalent to \fB\-\-dereference\-args\fR (\fB\-D\fR)
.TP
\fB\-h\fR, \fB\-\-human\-readable\fR
print sizes in human readable format (e.g., 1K 234M 2G)
.TP
\fB\-\-inodes\fR
list inode usage information instead of block usage
.TP
\fB\-k\fR
like \fB\-\-block\-size\fR=\fI\,1K\/\fR
.TP
\fB\-L\fR, \fB\-\-dereference\fR
dereference all symbolic links
.TP
\fB\-l\fR, \fB\-\-count\-links\fR
count sizes many times if hard linked
.TP
\fB\-m\fR
like \fB\-\-block\-size\fR=\fI\,1M\/\fR
.TP
\fB\-P\fR, \fB\-\-no\-dereference\fR
don't follow any symbolic links (this is the default)
.TP
\fB\-S\fR, \fB\-\-separate\-dirs\fR
for directories do not include size of subdirectories
.TP
\fB\-\-si\fR
like \fB\-h\fR, but use powers of 1000 not 1024
.TP
\fB\-s\fR, \fB\-\-summarize\fR
display only a total for each argument
.TP
\fB\-t\fR, \fB\-\-threshold\fR=\fI\,SIZE\/\fR
exclude entries smaller than SIZE if positive,
or entries greater than SIZE if negative
.TP
\fB\-\-time\fR
show time of the last modification of any file in the
directory, or any of its subdirectories
.TP
\fB\-\-time\fR=\fI\,WORD\/\fR
show time as WORD instead of modification time:
atime, access, use, ctime or status
.TP
\fB\-\-time\-style\fR=\fI\,STYLE\/\fR
show times using STYLE, which can be:
full\-iso, long\-iso, iso, or +FORMAT;
FORMAT is interpreted like in 'date'
.TP
\fB\-X\fR, \fB\-\-exclude\-from\fR=\fI\,FILE\/\fR
exclude files that match any pattern in FILE
.TP
\fB\-\-exclude\fR=\fI\,PATTERN\/\fR
exclude files that match PATTERN
.TP
\fB\-x\fR, \fB\-\-one\-file\-system\fR
skip directories on different file systems
.TP
\fB\-\-help\fR
display this help and exit
.TP
\fB\-\-version\fR
output version information and exit
.PP
Display values are in units of the first available SIZE from \fB\-\-block\-size\fR,
and the DU_BLOCK_SIZE, BLOCK_SIZE and BLOCKSIZE environment variables.
Otherwise, units default to 1024 bytes (or 512 if POSIXLY_CORRECT is set).
.PP
The SIZE argument is an integer and optional unit (example: 10K is 10*1024).
Units are K,M,G,T,P,E,Z,Y,R,Q (powers of 1024) or KB,MB,... (powers of 1000).
Binary prefixes can be used, too: KiB=K, MiB=M, and so on.
.SH PATTERNS
PATTERN is a shell pattern (not a regular expression).  The pattern
.B ?\&
matches any one character, whereas
.B *
matches any string (composed of zero, one or multiple characters).  For
example,
.B *.o
will match any files whose names end in
.BR .o .
Therefore, the command
.IP
.B du \-\-exclude=\(aq*.o\(aq
.PP
will skip all files and subdirectories ending in
.B .o
(including the file
.B .o
itself).
.SH AUTHOR
Written by Torbjorn Granlund, David MacKenzie, Paul Eggert,
and Jim Meyering.
.SH "REPORTING BUGS"
GNU coreutils online help: <https://www.gnu.org/software/coreutils/>
.br
Report any translation bugs to <https://translationproject.org/team/>
.SH COPYRIGHT
Copyright \(co 2024 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <https://gnu.org/licenses/gpl.html>.
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
Full documentation <https://www.gnu.org/software/coreutils/du>
.br
or available locally via: info \(aq(coreutils) du invocation\(aq


projects/busybox-1.37.0/coreutils/du.c
/* vi: set sw=4 ts=4: */
/*
 * Mini du implementation for busybox
 *
 * Copyright (C) 1999,2000,2001 by Lineo, inc. and John Beppu
 * Copyright (C) 1999,2000,2001 by John Beppu <beppu@codepoet.org>
 * Copyright (C) 2002  Edward Betts <edward@debian.org>
 *
 * Licensed under GPLv2 or later, see file LICENSE in this source tree.
 */
/* Mar 16, 2003      Manuel Novoa III   (mjn3@codepoet.org)
 *
 * Mostly rewritten for SUSv3 compliance and to fix bugs/defects.
 * 1) Added support for SUSv3 -a, -H, -L, gnu -c, and (busybox) -d options.
 *    The -d option allows setting of max depth (similar to gnu --max-depth).
 * 2) Fixed incorrect size calculations for links and directories, especially
 *    when errors occurred.  Calculates sizes should now match gnu du output.
 * 3) Added error checking of output.
 * 4) Fixed busybox bug #1284 involving long overflow with human_readable.
 */
//config:config DU
//config:	bool "du (6.5 kb)"
//config:	default y
//config:	help
//config:	du is used to report the amount of disk space used
//config:	for specified files.
//config:
//config:config FEATURE_DU_DEFAULT_BLOCKSIZE_1K
//config:	bool "Use default blocksize of 1024 bytes (else it's 512 bytes)"
//config:	default y
//config:	depends on DU

//applet:IF_DU(APPLET(du, BB_DIR_USR_BIN, BB_SUID_DROP))

//kbuild:lib-$(CONFIG_DU) += du.o

/* BB_AUDIT SUSv3 compliant (unless default blocksize set to 1k) */
/* http://www.opengroup.org/onlinepubs/007904975/utilities/du.html */

//usage:#define du_trivial_usage
//usage:       "[-aHLdclsx" IF_FEATURE_HUMAN_READABLE("hm") "k] [FILE]..."
//usage:#define du_full_usage "\n\n"
//usage:       "Summarize disk space used for FILEs (or directories)\n"
//usage:     "\n	-a	Show file sizes too"
//usage:     "\n	-b	Apparent size (including holes)"
//usage:     "\n	-L	Follow all symlinks"
//usage:     "\n	-H	Follow symlinks on command line"
//usage:     "\n	-d N	Limit output to directories (and files with -a) of depth < N"
//usage:     "\n	-c	Show grand total"
//usage:     "\n	-l	Count sizes many times if hard linked"
//usage:     "\n	-s	Display only a total for each argument"
//usage:     "\n	-x	Skip directories on different filesystems"
//usage:	IF_FEATURE_HUMAN_READABLE(
//usage:     "\n	-h	Sizes in human readable format (e.g., 1K 243M 2G)"
//usage:     "\n	-m	Sizes in megabytes"
//usage:	)
//usage:     "\n	-k	Sizes in kilobytes" IF_FEATURE_DU_DEFAULT_BLOCKSIZE_1K(" (default)")
//usage:	IF_NOT_FEATURE_DU_DEFAULT_BLOCKSIZE_1K(
//usage:     "\n		Default unit is 512 bytes"
//usage:	)
//usage:
//usage:#define du_example_usage
//usage:       "$ du\n"
//usage:       "16      ./CVS\n"
//usage:       "12      ./kernel-patches/CVS\n"
//usage:       "80      ./kernel-patches\n"
//usage:       "12      ./tests/CVS\n"
//usage:       "36      ./tests\n"
//usage:       "12      ./scripts/CVS\n"
//usage:       "16      ./scripts\n"
//usage:       "12      ./docs/CVS\n"
//usage:       "104     ./docs\n"
//usage:       "2417    .\n"

#include "libbb.h"
#include "common_bufsiz.h"

enum {
	OPT_a_files_too    = (1 << 0),
	OPT_H_follow_links = (1 << 1),
	OPT_k_kbytes       = (1 << 2),
	OPT_L_follow_links = (1 << 3),
	OPT_s_total_norecurse = (1 << 4),
	OPT_x_one_FS       = (1 << 5),
	OPT_d_maxdepth     = (1 << 6),
	OPT_l_hardlinks    = (1 << 7),
	OPT_c_total        = (1 << 8),
	OPT_b              = (1 << 9),
	OPT_h_for_humans   = (1 << 10),
	OPT_m_mbytes       = (1 << 11),
};

struct globals {
#if ENABLE_FEATURE_HUMAN_READABLE
	unsigned long disp_unit;
#else
	unsigned disp_k;
#endif
	int max_print_depth;
	bool status;
	int slink_depth;
	int du_depth;
	dev_t dir_dev;
} FIX_ALIASING;
#define G (*(struct globals*)bb_common_bufsiz1)
#define INIT_G() do { setup_common_bufsiz(); } while (0)


static void print(unsigned long long size, const char *filename)
{
	/* TODO - May not want to defer error checking here. */
#if ENABLE_FEATURE_HUMAN_READABLE
# if ENABLE_DESKTOP
	/* ~30 bytes of code for extra compat:
	 * coreutils' du rounds sizes up:
	 * for example,  1025k file is shown as "2" by du -m.
	 * We round to nearest if human-readable [too hard to fix],
	 * else (fixed scale such as -m), we round up. To that end,
	 * add yet another half of the unit before displaying:
	 */
	if (G.disp_unit)
		size += (G.disp_unit-1) / (unsigned)(512 * 2);
# endif
	printf("%s\t%s\n",
			/* size x 512 / G.disp_unit.
			 * If G.disp_unit == 0, show one fractional
			 * and use suffixes
			 */
			make_human_readable_str(size, (option_mask32 & OPT_b) ? 1 : 512, G.disp_unit),
			filename);
#else
	if (G.disp_k) {
		if (!(option_mask32 & OPT_b)) {
			size++;
			size >>= 1;
		} else {
			size >>= 10;
		}
	}
	printf("%llu\t%s\n", size, filename);
#endif
}

/* tiny recursive du */
static unsigned long long du(const char *filename)
{
	struct stat statbuf;
	unsigned long long sum;

	if (lstat(filename, &statbuf) != 0) {
		bb_simple_perror_msg(filename);
		G.status = EXIT_FAILURE;
		return 0;
	}

	if (option_mask32 & OPT_x_one_FS) {
		if (G.du_depth == 0) {
			G.dir_dev = statbuf.st_dev;
		} else if (G.dir_dev != statbuf.st_dev) {
			return 0;
		}
	}

	sum = ((option_mask32 & OPT_b) ? statbuf.st_size : statbuf.st_blocks);

	if (S_ISLNK(statbuf.st_mode)) {
		if (G.slink_depth > G.du_depth) { /* -H or -L */
			if (stat(filename, &statbuf) != 0) {
				bb_simple_perror_msg(filename);
				G.status = EXIT_FAILURE;
				return 0;
			}
			sum = ((option_mask32 & OPT_b) ? statbuf.st_size : statbuf.st_blocks);
			if (G.slink_depth == 1) {
				/* Convert -H to -L */
				G.slink_depth = INT_MAX;
			}
		}
	}

	if (!(option_mask32 & OPT_l_hardlinks)
	 && statbuf.st_nlink > 1
	) {
		/* Add files/directories with links only once */
		if (is_in_ino_dev_hashtable(&statbuf)) {
			return 0;
		}
		add_to_ino_dev_hashtable(&statbuf, NULL);
	}

	if (S_ISDIR(statbuf.st_mode)) {
		DIR *dir;
		struct dirent *entry;
		char *newfile;

		dir = warn_opendir(filename);
		if (!dir) {
			G.status = EXIT_FAILURE;
			return sum;
		}

		while ((entry = readdir(dir))) {
			newfile = concat_subpath_file(filename, entry->d_name);
			if (newfile == NULL)
				continue;
			++G.du_depth;
			sum += du(newfile);
			--G.du_depth;
			free(newfile);
		}
		closedir(dir);
	} else {
		if (!(option_mask32 & OPT_a_files_too) && G.du_depth != 0)
			return sum;
	}
	if (G.du_depth <= G.max_print_depth) {
		print(sum, filename);
	}
	return sum;
}

int du_main(int argc, char **argv) MAIN_EXTERNALLY_VISIBLE;
int du_main(int argc UNUSED_PARAM, char **argv)
{
	unsigned long long total;
	int slink_depth_save;
	unsigned opt;

	INIT_G();

#if ENABLE_FEATURE_HUMAN_READABLE
	IF_FEATURE_DU_DEFAULT_BLOCKSIZE_1K(G.disp_unit = 1024;)
	IF_NOT_FEATURE_DU_DEFAULT_BLOCKSIZE_1K(G.disp_unit = 512;)
	if (getenv("POSIXLY_CORRECT"))  /* TODO - a new libbb function? */
		G.disp_unit = 512;
#else
	IF_FEATURE_DU_DEFAULT_BLOCKSIZE_1K(G.disp_k = 1;)
	/* IF_NOT_FEATURE_DU_DEFAULT_BLOCKSIZE_1K(G.disp_k = 0;) - G is pre-zeroed */
#endif
	G.max_print_depth = INT_MAX;

	/* Note: SUSv3 specifies that -a and -s options cannot be used together
	 * in strictly conforming applications.  However, it also says that some
	 * du implementations may produce output when -a and -s are used together.
	 * gnu du exits with an error code in this case.  We choose to simply
	 * ignore -a.  This is consistent with -s being equivalent to -d 0.
	 */
#if ENABLE_FEATURE_HUMAN_READABLE
	opt = getopt32(argv, "^"
			"aHkLsxd:+lcbhm"
			"\0" "h-km:k-hm:m-hk:H-L:L-H:s-d:d-s",
			&G.max_print_depth
	);
	argv += optind;
	if (opt & OPT_b) {
		G.disp_unit = 1;
	}
	if (opt & OPT_h_for_humans) {
		G.disp_unit = 0;
	}
	if (opt & OPT_m_mbytes) {
		G.disp_unit = 1024*1024;
	}
	if (opt & OPT_k_kbytes) {
		G.disp_unit = 1024;
	}
#else
	opt = getopt32(argv, "^"
			"aHkLsxd:+lcb"
			"\0" "H-L:L-H:s-d:d-s",
			&G.max_print_depth
	);
	argv += optind;
# if !ENABLE_FEATURE_DU_DEFAULT_BLOCKSIZE_1K
	if (opt & OPT_k_kbytes) {
		G.disp_k = 1;
	}
# endif
#endif
	if (opt & OPT_H_follow_links) {
		G.slink_depth = 1;
	}
	if (opt & OPT_L_follow_links) {
		G.slink_depth = INT_MAX;
	}
	if (opt & OPT_s_total_norecurse) {
		G.max_print_depth = 0;
	}

	/* go through remaining args (if any) */
	if (!*argv) {
		*--argv = (char*)".";
		if (G.slink_depth == 1) {
			G.slink_depth = 0;
		}
	}

	slink_depth_save = G.slink_depth;
	total = 0;
	do {
		total += du(*argv);
		G.slink_depth = slink_depth_save;
	} while (*++argv);

	if (ENABLE_FEATURE_CLEAN_UP)
		reset_ino_dev_hashtable();
	if (opt & OPT_c_total)
		print(total, "total");

	fflush_stdout_and_exit(G.status);
}
